kubeflow/trainer/backends/kubernetes/backend.py

```diff
@@ -518,12 +518,39 @@ def __get_trainjob_from_cr(
             if not pod_list:
                 return trainjob
 
+            # Group Pods by their role to handle Pod restarts/recreations. select only the most
+            # recently created Pod for each component to show users the current state.
+            pod_groups: dict[str, list] = {}
             for pod in pod_list.items:
                 # Pod must have labels to detect the TrainJob step.
                 # Every Pod always has a single TrainJob step.
                 if not (pod.metadata and pod.metadata.name and pod.metadata.labels and pod.spec):
                     raise Exception(f"TrainJob Pod is invalid: {pod}")
 
+                # Create unique key for each TrainJob component:
+                # - For initializers: use the role name (dataset-initializer, model-initializer)
+                # - For training nodes: use role + job index (node-0, node-1, launcher-0, etc.)
+                role = pod.metadata.labels[constants.JOBSET_RJOB_NAME_LABEL]
+                if role in {constants.LAUNCHER, constants.NODE}:
+                    job_index = pod.metadata.labels.get(constants.JOB_INDEX_LABEL, "0")
+                    key = f"{role}-{job_index}"
+                else:
+                    key = role
+
+                if key not in pod_groups:
+                    pod_groups[key] = []
+                pod_groups[key].append(pod)
+
+            # Select the most recently created Pod from each group.
```

Comment:
I think to make it more robust we could select the pod based on the status as well as the timestamp something like this https://github.com/Fiona-Waters/kubeflow-sdk/commit/b48277f78fdb46e1dbdd8667fbc239ff044b6828
wdyt?
It will return a pod that actually reflects the true state of each TrainJob component, rather than the newest pod.